src/soep_preparation/final_dataset.py

```diff
@@ -0,0 +1,204 @@
+
+    Returns:
+        The dataset with specified variables and survey years.
+
+    Raises:
+        TypeError: If the input types are not as expected.
+        ValueError: If variables or variable mapping are empty or
+        contain faulty variables.
+        Or if survey years out of survey range.
+        Or if inadequate merging behavior.
+
+    Notes:
+        `modules` contains the cleaned and combined modules with the
+         variables of interest.
+        `variable_to_metadata` is created automatically by the pipeline,
+        it can be accessed and provided to the function at
+        `src/soep_preparation/create_metadata/variable_to_metadata_mapping.yaml`.
+        `variables` contains the variables
+        created, renamed, and derived from the raw SOEP data files
+        that will be part of the merged dataset.
+        `survey_years` are those years the final dataset will contain variables for.
+        To receive data for just one year (e.g. `2025`) specify `survey_years=[2025]`.
+
+    Examples:
+        For an example see `task_example.py`.
+    """
+    _error_handling(
+        modules=modules,
+        variable_to_metadata=variable_to_metadata,
+        variables=variables,
+        survey_years=survey_years,
+    )
+    harmonized_variables = _harmonize_variables(variables)
+
+    dataset_merging_information = _get_sorted_dataset_merging_information(
+        modules=modules,
+        variable_to_metadata=variable_to_metadata,
+        variables=harmonized_variables,
+        survey_years=survey_years,
+    )
+
+    return _merge_data(
+        merging_information=dataset_merging_information,
+    )
+
+
+def _error_handling(
+    modules: dict[str, pd.DataFrame],
+    variable_to_metadata: dict[str, dict],
+    variables: list[str],
+    survey_years: list[int],
+) -> None:
+    fail_if_input_has_invalid_type(
+        input_=variable_to_metadata,
+        expected_dtypes=["dict", "PNode", "PProvisionalNode"],
+    )
+    fail_if_input_has_invalid_type(input_=variables, expected_dtypes=["list"])
+    fail_if_input_has_invalid_type(
+        input_=survey_years, expected_dtypes=("list", "None")
+    )
+    fail_if_empty(input_=variable_to_metadata, name="variable_to_metadata")
+    fail_if_empty(variables, name="variables")
+
+    # TODO (@hmgaudecker): we need at-least one module with the variable  # noqa: TD003
+    # `survey_year` to check for valid survey years
+    valid_survey_years = modules["pl"]["survey_year"].unique().tolist()
+    _fail_if_survey_years_not_valid(
+        survey_years=survey_years,
+        valid_survey_years=valid_survey_years,
+    )
+    _fail_if_invalid_variable(
+        variables=variables, variable_to_metadata=variable_to_metadata
+    )
+
+
+def _fail_if_invalid_variable(
+    variables: list[str],
+    variable_to_metadata: dict[str, dict],
+) -> None:
+    for variable in variables:
+        if variable not in variable_to_metadata and variable not in ID_VARIABLES:
```

Comment:
Is this not redundant? How could we have a POTENTIAL_INDEX_VARIABLE that is not in the metadata?